{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "11429817-04e9-4e05-8fc0-d3e59bf6dda5",
   "metadata": {},
   "source": [
    "# This step will run EnDecon.\n",
    "\n",
    "### Note this is R code runnning in jupyter!\n",
    "### Use Kernel: R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9fef8e11-db62-448d-87b8-f3633dcd582e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(EnDecon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "331f5f55-c870-4a6a-a7f7-bad2fd3d19db",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Optional, only if you want to double check your library has been installed\n",
    "library(devtools)\n",
    "library (curl)\n",
    "library (RcppGSL)\n",
    "library(Rfast)\n",
    "library(fastmatrix)\n",
    "library(BiocManager)\n",
    "library(Seurat)\n",
    "library(CARD)\n",
    "library(reticulate)\n",
    "library(TOAST)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc29e4ca-2194-4f85-a7d8-f6fcf3efdfa1",
   "metadata": {},
   "source": [
    "## Step 1 Assign ref_id for EnDecon input files (synthetic files and reference files)\n",
    "Here you will assign with sim_id file to analyze."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdc63d76-ca8e-46c2-a85f-8746caa26e7f",
   "metadata": {},
   "source": [
    "#### Input file for scenario 3 as example, you need to adjust ref_id and sim_id for Scenario 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d8556bb2-03e4-438d-886e-53000ee4b17c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_id = '6702e4f7a944d6095c2c2d1496311866'\n",
    "\n",
    "#sim_id = '49db3a23106485d79a11c5f9ec2be108'\n",
    "#sim_id_list <- c('49db3a23106485d79a11c5f9ec2be108','4555a63ca602ee279f81713e3f30adfb','24bc51ffa07c227a8aaab7f616fa7166',)\n",
    "sim_id_list <- c(\n",
    "    '43885de192f6b406821c09b949f8e5ea_mixlevel2',\n",
    "    '43885de192f6b406821c09b949f8e5ea_mixlevel0',\n",
    "    '43885de192f6b406821c09b949f8e5ea_mixlevel1',\n",
    "    '43885de192f6b406821c09b949f8e5ea_mixlevel3'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7bd1dcc-afa2-419a-849a-6698b3e600b1",
   "metadata": {},
   "source": [
    "## Step 2 run EnDecon\n",
    "Here you need to adjust the code to choose which deconv method to run. I suggest run cell2location seperately, since it will take a long time. In the case of kernel died during the run, you still have output of other methods saved. Usually takes 30-50 minutes for each dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8a2e09d1-e028-464e-939e-2e8ca2589ac0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Now run EnDecon for sim_id:\"\n",
      "[1] \"43885de192f6b406821c09b949f8e5ea_mixlevel2\"\n",
      "Execute CARD analysis....\n",
      "## QC on scRNASeq dataset! ...\n",
      "## QC on spatially-resolved dataset! ...\n",
      "## create reference matrix from scRNASeq...\n",
      "## Select Informative Genes! ...\n",
      "## Deconvolution Starts! ...\n",
      "## Deconvolution Finish! ...\n",
      "Run time for CARD:  0.1945684 min \n",
      "Execute DeconRNASeq analysis....\n",
      "svd calculated PCA\n",
      "Importance of component(s):\n",
      "                 PC1     PC2     PC3     PC4     PC5     PC6     PC7     PC8\n",
      "R2            0.7141 0.04235 0.02175 0.02061 0.01966 0.01805 0.01326 0.01161\n",
      "Cumulative R2 0.7141 0.75644 0.77819 0.79880 0.81846 0.83651 0.84977 0.86138\n",
      "                  PC9\n",
      "R2            0.01103\n",
      "Cumulative R2 0.87241\n",
      "\n",
      " Attention: the number of pure cell types = 9  defined in the signature matrix;\n",
      "\n",
      " PCA results indicate that the number of cell types in the mixtures = 1 \n",
      "Run time for DeconRNASeq:  0.1331774 min \n",
      "Execute DestVI analysis....\n",
      "Run time for DestVI:  13.63232 min \n",
      "Execute RCTD analysis....\n",
      "\n",
      "                                    B cell \n",
      "                                      1179 \n",
      "           CD4-positive, alpha-beta T cell \n",
      "                                      4572 \n",
      "           CD8-positive, alpha-beta T cell \n",
      "                                      5000 \n",
      "alveolar capillary type 1 endothelial cell \n",
      "                                       138 \n",
      "alveolar capillary type 2 endothelial cell \n",
      "                                        74 \n",
      "                   epithelial cell of lung \n",
      "                                        23 \n",
      "                        fibroblast of lung \n",
      "                                      2568 \n",
      "       tracheobronchial smooth muscle cell \n",
      "                                       216 \n",
      "                         type I pneumocyte \n",
      "                                        94 \n",
      "Run time for RCTD:  12.85761 min \n",
      "Execute SCDC analysis....\n",
      "Run time for SCDC:  4.146525 min \n",
      "Execute SPOTlight analysis....\n",
      "[1] \"Preparing Gene set\"\n",
      "[1] \"Normalizing count matrix\"\n",
      "[1] \"Seeding initial matrices\"\n",
      "[1] \"Training...\"\n",
      "[1] \"Time to train NMF model was 10.51mins\"\n",
      "[1] \"Deconvoluting spots\"\n",
      "  |======================================================================| 100%\n",
      "Run time for SPOTlight:  14.76797 min \n",
      "Execute STdeconvolve analysis....\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Removing 397 genes present in 100% or more of pixels...\n",
      "\n",
      "8677 genes remaining...\n",
      "\n",
      "Removing 668 genes present in 5% or less of pixels...\n",
      "\n",
      "8009 genes remaining...\n",
      "\n",
      "Restricting to overdispersed genes with alpha = 0.05...\n",
      "\n",
      "Calculating variance fit ...\n",
      "\n",
      "Using gam with k=5...\n",
      "\n",
      "1806 overdispersed genes ... \n",
      "\n",
      " Using top 1000 overdispersed genes.\n",
      "\n",
      "\n",
      "NOTE: using type='b' and comparing betas where the cell-types are\n",
      "            the rows (cell-types x genes)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Run time for STdeconvolve:  1.9073 min \n",
      "        CARD  DeconRNASeq       DestVI         RCTD         SCDC    SPOTlight \n",
      "   0.1945684    0.1331774   13.6323207   12.8576080    4.1465253   14.7679732 \n",
      "STdeconvolve \n",
      "   1.9073003 \n",
      "[1] 1\n",
      "[1] 2\n",
      "[1] 3\n",
      "[1] 4\n",
      "[1] 5\n",
      "[1] 6\n",
      "[1] 7\n"
     ]
    }
   ],
   "source": [
    "for (sim_id in sim_id_list) {\n",
    "print(\"Now run EnDecon for sim_id:\")   \n",
    "print(sim_id)\n",
    "    \n",
    "# file location for synthetic data (processed for CARD/EnDecon)\n",
    "spatial_path = paste0(\"/home/oneai/data/EnDecon_input/\", ref_id, \"/\", \"simulated_SRT_dataset_\", sim_id, \"/\")\n",
    "#reference data path (storing processed sc_meta.csv and sc_count.csv)\n",
    "sc_path = paste0(\"/home/oneai/data/EnDecon_input/\", ref_id, \"/\")\n",
    "# Output path for EnDecon\n",
    "output_path = \"/home/oneai/data/EnDecon_output/\"\n",
    "##### path on ubuntu platform on our computer\n",
    "python_env <- \"~/envs/EnDecon_GPU_New/bin/python\" #JX's path adjusted  \n",
    "    \n",
    "\n",
    "    \n",
    "#STEP1 single cell transcriptomics data     \n",
    "# This step takes longer time. Read in the sc reference data\n",
    "sc_count_syn <- read.csv(paste0(sc_path, \"sc_count.csv\"), header = TRUE)\n",
    "sc_count_syn <- as.matrix(sc_count_syn)\n",
    "# Use first column (gene name as row name)\n",
    "row.names(sc_count_syn) <- sc_count_syn[, 1] \n",
    "# Remove the first column from the data (optional)\n",
    "sc_count_syn <- sc_count_syn[, -1]\n",
    "# Transfer the value to integer before sparseMatrix\n",
    "mode(sc_count_syn) = \"integer\"\n",
    "sc_count_syn[1:6,0:10]\n",
    "\n",
    "#### STEP2 single cell RNAseq ((scRNA-seq)) data, \n",
    "# read the cellID and cell_type, store as named character vector (same format as sc_meta). \n",
    "# Use sc_meta_syn column \"cell_id\" as the unique names or identifiers (CIDs), use \"cell_type\" coumn as corresponding values or annotations.\n",
    "sc_meta_syn <- read.csv(paste0(sc_path, \"sc_meta.csv\"), header = TRUE)\n",
    "sc_meta_syn <- setNames(sc_meta_syn$cell_type, sc_meta_syn$cellID)\n",
    "sc_meta_syn[1:3]    \n",
    "\n",
    "#### STEP3 spatial transcriptomics count data    \n",
    "spatial_count_syn <- read.csv(paste0(spatial_path, \"spatial_count.csv\"), header = TRUE)\n",
    "spatial_count_syn <- as.matrix(spatial_count_syn)\n",
    "# Use first column (gene name) as row name\n",
    "row.names(spatial_count_syn) <- spatial_count_syn[, 1] \n",
    "# Remove the first column from the data (optional)\n",
    "spatial_count_syn <- spatial_count_syn[, -1]\n",
    "# Transfer the value to integer \n",
    "mode(spatial_count_syn) = \"integer\"   \n",
    "spatial_count_syn[1:5,1:5]\n",
    "\n",
    "#### STEP4 spatial location\n",
    "spatial_location_syn <- read.csv(paste0(spatial_path, \"spatial_location.csv\"), header = TRUE)\n",
    "# Use first column (gene name as row name)\n",
    "row.names(spatial_location_syn) <- spatial_location_syn[, 1] \n",
    "# Remove the first column from the data (optional)\n",
    "spatial_location_syn <- spatial_location_syn[, -1]\n",
    "# Revert y axis!! This y is the real x-y axis. R and python are reverted!!\"\n",
    "spatial_location_syn$y <- rev (spatial_location_syn$y)\n",
    "spatial_location_syn[1:4,]\n",
    "\n",
    "\n",
    "    \n",
    "#### Run EnDecon for all method, for synthetic data\n",
    "Results.dec.mouse <- EnDecon_individual_methods(sc_exp = sc_count_syn,\n",
    "                                                sc_label = sc_meta_syn, spot_exp = spatial_count_syn,\n",
    "                                                spot_loc = spatial_location_syn, python_env = python_env,\n",
    "                                                use_gpu = TRUE,gene_det_in_min_cells_per = 0.01,\n",
    "                                                RCTD.CELL_MIN_INSTANCE = 5, saving_results = FALSE, \n",
    "                                                \n",
    "                                                SCDC = TRUE, RCTD = TRUE, DeconRNASeq = TRUE,\n",
    "                                                DestVI = TRUE, DWLS = TRUE, SPOTlight = TRUE, SpatialDWLS = TRUE,\n",
    "                                                cell2location = TRUE, CARD = TRUE, STdeconvolve = TRUE, \n",
    "                                                \n",
    "                                                Stereoscope = FALSE, MuSiC = FALSE, #these two didn't work \n",
    "                                                \n",
    "                                                cell2location.sc_max_epoches = 800, cell2location.sc_lr = 0.002, cell2location.st_N_cells_per_location = 15,\n",
    "                                                cell2location.st_detection_alpha = 200.00, cell2location.st_max_epoches = 8000\n",
    "                                               )    \n",
    "\n",
    "\n",
    "\n",
    "print(Results.dec.mouse[[2]])\n",
    "\n",
    "\n",
    "#### Save RData and Runtime into folder: /home/oneai/data/EnDecon_output/RData\n",
    "rdata = Results.dec.mouse[[1]]\n",
    "runtime = Results.dec.mouse[[2]]\n",
    "df <- data.frame(Method = names(runtime), Runtime = as.character(runtime), stringsAsFactors = FALSE)\n",
    "save(rdata, file = paste0(output_path, \"/RData/simulated_SRT_dataset_\", sim_id, \"_Results_Deconv.RData\"))\n",
    "write.csv(df, file = paste0(output_path,\"/RData/simulated_SRT_dataset_\",  sim_id, \"_Runtime.csv\"), row.names = FALSE)       \n",
    "                                                             \n",
    "\n",
    "\n",
    "#### Save method-prop.csv\n",
    "for (i in 1:length(Results.dec.mouse[[1]])){\n",
    "    print (i)\n",
    "    write.csv(Results.dec.mouse[[1]][i], file = paste0(output_path, \"simulated_SRT_dataset_\", sim_id, \"_\", names(Results.dec.mouse[[1]][i]), \"_prop.csv\"), row.names = TRUE)\n",
    "    }    \n",
    "    \n",
    "    \n",
    "    \n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dee8e0d6-576c-448c-b143-7b356521c373",
   "metadata": {},
   "source": [
    "## Optional "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9de313f-7001-41c5-b612-d42c71dde186",
   "metadata": {},
   "outputs": [],
   "source": [
    "### If Only Run 3 fast R method to quick check\n",
    "Results.dec.mouse <- EnDecon_individual_methods(sc_exp = sc_count_syn,\n",
    "                                                sc_label = sc_meta_syn, spot_exp = spatial_count_syn,\n",
    "                                                spot_loc = spatial_location_syn, python_env = python_env,\n",
    "                                                use_gpu = TRUE,gene_det_in_min_cells_per = 0.01,\n",
    "                                                RCTD.CELL_MIN_INSTANCE = 5, saving_results = TRUE, \n",
    "                                                \n",
    "                                                SCDC = FALSE, RCTD = FALSE, MuSiC = FALSE, DeconRNASeq = TRUE,\n",
    "                                                DestVI = FALSE, DWLS = FALSE, SPOTlight = FALSE, SpatialDWLS = FALSE,\n",
    "                                                Stereoscope = FALSE, cell2location = FALSE, CARD = TRUE, STdeconvolve = TRUE,\n",
    "                                                \n",
    "                                                #cell2location.sc_max_epoches = 100, cell2location.sc_lr = 0.002, cell2location.st_N_cells_per_location = 30,\n",
    "                                                #cell2location.st_detection_alpha = 200.00, cell2location.st_max_epoches = 1000\n",
    "                                               )   "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
