{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "09405f94-6732-4cdf-96aa-4d730dce9d01",
   "metadata": {},
   "source": [
    "# This step will run EnDecon.\n",
    "\n",
    "### Note this is R code runnning in jupyter!\n",
    "### Use Kernel: R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9fef8e11-db62-448d-87b8-f3633dcd582e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(EnDecon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "331f5f55-c870-4a6a-a7f7-bad2fd3d19db",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: usethis\n",
      "\n",
      "Using libcurl 7.68.0 with OpenSSL/1.1.1f\n",
      "\n",
      "Loading required package: Rcpp\n",
      "\n",
      "\n",
      "Attaching package: ‘Rcpp’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:RcppGSL’:\n",
      "\n",
      "    LdFlags\n",
      "\n",
      "\n",
      "Loading required package: RcppZiggurat\n",
      "\n",
      "\n",
      "Rfast:  2.0.8\n",
      "\n",
      " ___ __ __ __ __    __ __ __ __ __ _             _               __ __ __ __ __     __ __ __ __ __ __   \n",
      "|  __ __ __ __  |  |  __ __ __ __ _/            / \\             |  __ __ __ __ /   /__ __ _   _ __ __\\  \n",
      "| |           | |  | |                         / _ \\            | |                        / /          \n",
      "| |           | |  | |                        / / \\ \\           | |                       / /          \n",
      "| |           | |  | |                       / /   \\ \\          | |                      / /          \n",
      "| |__ __ __ __| |  | |__ __ __ __           / /     \\ \\         | |__ __ __ __ _        / /__/\\          \n",
      "|    __ __ __ __|  |  __ __ __ __|         / /__ _ __\\ \\        |_ __ __ __ _   |      / ___  /           \n",
      "|   \\              | |                    / _ _ _ _ _ _ \\                     | |      \\/  / /       \n",
      "| |\\ \\             | |                   / /           \\ \\                    | |         / /          \n",
      "| | \\ \\            | |                  / /             \\ \\                   | |        / /          \n",
      "| |  \\ \\           | |                 / /               \\ \\                  | |       / /          \n",
      "| |   \\ \\__ __ _   | |                / /                 \\ \\     _ __ __ __ _| |      / /          \n",
      "|_|    \\__ __ __\\  |_|               /_/                   \\_\\   /_ __ __ __ ___|      \\/             team\n",
      "\n",
      "\n",
      "Attaching package: ‘Rfast’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocManager’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:devtools’:\n",
      "\n",
      "    install\n",
      "\n",
      "\n",
      "Attaching SeuratObject\n",
      "\n",
      "Loading required package: EpiDISH\n",
      "\n",
      "Loading required package: limma\n",
      "\n",
      "\n",
      "Attaching package: ‘limma’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "Loading required package: nnls\n",
      "\n",
      "Loading required package: quadprog\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Optional, only if you want to double check your library has been installed\n",
    "library(devtools)\n",
    "library (curl)\n",
    "library (RcppGSL)\n",
    "library(Rfast)\n",
    "library(fastmatrix)\n",
    "library(BiocManager)\n",
    "library(Seurat)\n",
    "library(CARD)\n",
    "library(reticulate)\n",
    "library(TOAST)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7511157-d475-4868-b18d-a358932e2da4",
   "metadata": {},
   "source": [
    "## Step 1 Input file\n",
    "### adjust the ref_id, tail, sim_id depending on your scenarios"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57774f54-17c4-4ac9-93b9-d11e1c8e3f0f",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Input files for Scenario 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d8556bb2-03e4-438d-886e-53000ee4b17c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_id = 'feff311061feb64e82aaf93071ed1d86'\n",
    "\n",
    "# In the case of haveing multiple reference data, attaching tail for reference, otherwise, tail is empty\n",
    "# tail_of_ref=['']  # If there is no tail for reference adata. eg. Reference_for_deconv.h5ad\n",
    "tail_list <- c(\n",
    "    '_Lafyatis_Rojas_2019',\n",
    "    '_Teichmann_Meyer_2019',\n",
    "    '_Krasnow_2020',\n",
    "    '_Meyer_2019',\n",
    "    '_Banovich_Kropski_2020'\n",
    "    )\n",
    "\n",
    "#sim_id = '49db3a23106485d79a11c5f9ec2be108'\n",
    "#sim_id_list <- c('49db3a23106485d79a11c5f9ec2be108','4555a63ca602ee279f81713e3f30adfb','24bc51ffa07c227a8aaab7f616fa7166',)\n",
    "sim_id_list <- c('b78ba209152da0bcf71ec2356d867145')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2b8de4f-14ec-47c1-baf0-cb0841db424b",
   "metadata": {},
   "source": [
    "#### Input file for Scenario 1 or 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9b1f75f0-d5e7-41cc-bcc1-242573ca562a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For scenario 3:\n",
    "ref_id = '6702e4f7a944d6095c2c2d1496311866'\n",
    "sim_id_list <- c(\n",
    "    '43885de192f6b406821c09b949f8e5ea_mixlevel2',\n",
    "    '43885de192f6b406821c09b949f8e5ea_mixlevel0',\n",
    "    '43885de192f6b406821c09b949f8e5ea_mixlevel1',\n",
    "    '43885de192f6b406821c09b949f8e5ea_mixlevel3'\n",
    ")\n",
    "## You need to change red_id and sim_id_list for scenario 1 \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "803bff48-13fe-4410-b117-4a112a755ce2",
   "metadata": {},
   "source": [
    "## Step 2 run EnDecon\n",
    "Here you need to adjust the code to choose which deconv method to run. I suggest run cell2location seperately, since it will take a long time. In the case of kernel died during the run, you still have output of other methods saved. Usually takes 30-50 minutes for each dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a2e09d1-e028-464e-939e-2e8ca2589ac0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for (tail_of_ref in tail_list) {\n",
    "print(\"NEW REFERENCE !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!:\") \n",
    "print(\"Now run EnDecon for ref_id:\")   \n",
    "    \n",
    "print(tail_of_ref)\n",
    "    \n",
    "for (sim_id in sim_id_list) {\n",
    "print(\"Now run EnDecon for sim_id:\")   \n",
    "print(sim_id)\n",
    "    \n",
    "# file location for synthetic data (processed for CARD/EnDecon)\n",
    "spatial_path = paste0(\"/home/oneai/data/EnDecon_input/\", ref_id, \"/\", \"simulated_SRT_dataset_\", sim_id, \"/\")\n",
    "#reference data path (storing processed sc_meta.csv and sc_count.csv)\n",
    "sc_path = paste0(\"/home/oneai/data/EnDecon_input/\", ref_id, \"/\")\n",
    "# Output path for EnDecon\n",
    "output_path = \"/home/oneai/data/EnDecon_output/\"\n",
    "##### path on ubuntu platform on our computer\n",
    "python_env <- \"~/envs/EnDecon_GPU_New/bin/python\" #JX's path adjusted  \n",
    "    \n",
    "\n",
    "    \n",
    "#STEP1 single cell transcriptomics data     \n",
    "# This step takes longer time. Read in the sc reference data\n",
    "sc_count_syn <- read.csv(paste0(sc_path, \"sc_count\", tail_of_ref,\".csv\"), header = TRUE)\n",
    "sc_count_syn <- as.matrix(sc_count_syn)\n",
    "# Use first column (gene name as row name)\n",
    "row.names(sc_count_syn) <- sc_count_syn[, 1] \n",
    "# Remove the first column from the data (optional)\n",
    "sc_count_syn <- sc_count_syn[, -1]\n",
    "# Transfer the value to integer before sparseMatrix\n",
    "mode(sc_count_syn) = \"integer\"   #It was \"numeric\" for set 1 dataset, for mix reference or atlas, it is not integer, has to be change to integer\n",
    "sc_count_syn[1:6,0:10]\n",
    "\n",
    "#### STEP2 single cell RNAseq ((scRNA-seq)) data, \n",
    "# read the cellID and cell_type, store as named character vector (same format as sc_meta). \n",
    "# Use sc_meta_syn column \"cell_id\" as the unique names or identifiers (CIDs), use \"cell_type\" coumn as corresponding values or annotations.\n",
    "sc_meta_syn <- read.csv(paste0(sc_path, \"sc_meta\", tail_of_ref,\".csv\"), header = TRUE)\n",
    "sc_meta_syn <- setNames(sc_meta_syn$cell_type, sc_meta_syn$cellID)\n",
    "sc_meta_syn[1:3]    \n",
    "\n",
    "#### STEP3 spatial transcriptomics count data    \n",
    "spatial_count_syn <- read.csv(paste0(spatial_path, \"spatial_count.csv\"), header = TRUE)\n",
    "spatial_count_syn <- as.matrix(spatial_count_syn)\n",
    "# Use first column (gene name) as row name\n",
    "row.names(spatial_count_syn) <- spatial_count_syn[, 1] \n",
    "# Remove the first column from the data (optional)\n",
    "spatial_count_syn <- spatial_count_syn[, -1]\n",
    "# Transfer the value to integer \n",
    "mode(spatial_count_syn) = \"integer\"   \n",
    "spatial_count_syn[1:5,1:5]\n",
    "\n",
    "#### STEP4 spatial location\n",
    "spatial_location_syn <- read.csv(paste0(spatial_path, \"spatial_location.csv\"), header = TRUE)\n",
    "# Use first column (gene name as row name)\n",
    "row.names(spatial_location_syn) <- spatial_location_syn[, 1] \n",
    "# Remove the first column from the data (optional)\n",
    "spatial_location_syn <- spatial_location_syn[, -1]\n",
    "# Revert y axis!! This y is the real x-y axis. R and python are reverted!!\"\n",
    "spatial_location_syn$y <- rev (spatial_location_syn$y)\n",
    "spatial_location_syn[1:4,]\n",
    "\n",
    "    \n",
    "    \n",
    "################################################################Make sure get overlapped genes before run!!!\n",
    "# Get the overlapping row names\n",
    "common_row_names <- intersect(rownames(sc_count_syn), rownames(spatial_count_syn))\n",
    "# Keep only the common rows in sc_count_syn and spatial_count\n",
    "sc_count_syn <- sc_count_syn[common_row_names, ]\n",
    "spatial_count_syn <- spatial_count_syn[common_row_names, ]        \n",
    "    \n",
    "   \n",
    "    \n",
    "#### Run EnDecon for all method, for synthetic data, not Cell2location due to error!!\n",
    "Results.dec.mouse <- EnDecon_individual_methods(sc_exp = sc_count_syn,\n",
    "                                                sc_label = sc_meta_syn, spot_exp = spatial_count_syn,\n",
    "                                                spot_loc = spatial_location_syn, python_env = python_env,\n",
    "                                                use_gpu = TRUE,gene_det_in_min_cells_per = 0.01,\n",
    "                                                RCTD.CELL_MIN_INSTANCE = 5, saving_results = FALSE, \n",
    "                                                \n",
    "                                                #SCDC = TRUE, RCTD = TRUE, DeconRNASeq = TRUE,\n",
    "                                                #DestVI = TRUE, DWLS = FALSE, SPOTlight = TRUE, SpatialDWLS = FALSE,\n",
    "                                                #cell2location = FALSE, CARD = TRUE, STdeconvolve = TRUE,\n",
    "                                                \n",
    "                                                SCDC = TRUE, RCTD = TRUE, DeconRNASeq = TRUE,\n",
    "                                                DestVI = TRUE, DWLS = TRUE, SPOTlight = TRUE, SpatialDWLS = TRUE,\n",
    "                                                cell2location = TRUE, CARD = TRUE, STdeconvolve = TRUE,\n",
    "                                                \n",
    "                                                \n",
    "                                                Stereoscope = FALSE, MuSiC = FALSE, #These two methods wont work\n",
    "                                                \n",
    "                                                cell2location.sc_max_epoches = 800, cell2location.sc_lr = 0.002, cell2location.st_N_cells_per_location = 10,\n",
    "                                                cell2location.st_detection_alpha = 200.00, cell2location.st_max_epoches = 8000\n",
    "                                               )    \n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "print(Results.dec.mouse[[2]])\n",
    "\n",
    "\n",
    "#### Save RData and Runtime into folder: /home/oneai/data/EnDecon_output/RData\n",
    "rdata = Results.dec.mouse[[1]]\n",
    "runtime = Results.dec.mouse[[2]]\n",
    "df <- data.frame(Method = names(runtime), Runtime = as.character(runtime), stringsAsFactors = FALSE)\n",
    "save(rdata, file = paste0(output_path, \"/RData/simulated_SRT_dataset_\", sim_id, tail_of_ref, \"_Results_Deconv_C2L.RData\"))\n",
    "write.csv(df, file = paste0(output_path,\"/RData/simulated_SRT_dataset_\",  sim_id, tail_of_ref, \"_Runtime_C2L.csv\"), row.names = FALSE)       \n",
    "                                                             \n",
    "\n",
    "\n",
    "#### Save method-prop.csv\n",
    "for (i in 1:length(Results.dec.mouse[[1]])){\n",
    "    print (i)\n",
    "    write.csv(Results.dec.mouse[[1]][i], file = paste0(output_path, \"simulated_SRT_dataset_\", sim_id, tail_of_ref, \"_\", names(Results.dec.mouse[[1]][i]), \"_prop.csv\"), row.names = TRUE)\n",
    "    }    \n",
    "    \n",
    "    \n",
    "    \n",
    "}\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
